### PR TITLE
Issue-9: Add initial support for Validation

### LINKED ISSUE #9: Add initial support for Validation
Requires an implementation of `Tuple`.

Must include the following:

* A `Validation` type.
* The following static constructors:
    * `Failure a`
    * `Success b`
* The following static methods:
    * `Concat :: Validation a b → Validation a b → Validation a b`: Concatenates two `Failure` values of `a` together, replace a `Success` with the `Failure`; otherwise, take the first `Success.
    * `Failures :: [Validation a b] → [a]`: Extracts from a list of `Validation` all of the `Failure` elements in extracted order.
    * `FromFailure :: a → Validation a b → a`: Extracts the value of a `Failure`; otherwise, returns the `defaultValue`.
    * `FromSuccess :: b → Validation a b → b`: Extracts the value of a `Success`; otherwise, returns the `defaultValue`.
    * `IsFailure :: Validation a b → Boolean`: Determines whether or not the argument is a `Failure`.
    * `IsSuccess :: Validation a b → Boolean`: Determines whether or not the argument is a `Success`.
    * `PartitionValidations :: [Validation a b] → ([a], [b])`: Partitions a list of `Validation` into two lists. All `Failure` elements are extracted, in order, to the first position of the output. Similarly for the `Success` elements in the second position.
    * `Successes :: [Validation a b] → [b]`: Extracts from a list of `Validation` all of the `Success` elements in extracted order.
    * `Validate :: (b → Boolean) → a → b → Validation a b`: Validates a value `b` and returns a `Success` of `b` if the `predicate` returns `true`; otherwise, a `Failure` of `a`.
    * `ValidationMap :: ([a] → c) → (b → c) → Validation a b → c`: If the value is `Failure a`, apply the first function to `[a]`; otherwise, apply the second function to `b`.

### PR DESCRIPTION
**Pull request by jlmorgan**
- [x] Squash commits before merging.

closes #9
SemVer:Minor
enhancement

### CODE REVIEW

File: docs/Data/Validation/instance/Equals.md [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+* `B`: The underlying success type.
+
+## Returns
+
+* `(boolean)`: `true` for equality; otherwise, `false`.
+
+## Examples
+
+```csharp
+const failure = Validation.Failure<string, string>("a");
```
Comment (JeremiahSanders) [inline docs/Data/Validation/instance/Equals.md]:
```suggestion
var failure = Validation.Failure<string, string>("a");
```

File: docs/Data/Validation/instance/Equals.md [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+const success = Validation.Success<string, string>("a");
```
Comment (JeremiahSanders) [inline docs/Data/Validation/instance/Equals.md]:
```suggestion
var success = Validation.Success<string, string>("a");
```

File: docs/Data/Validation/instance/Equals.md (hunk 1 above)
Comment (jlmorgan) [reply docs/Data/Validation/instance/Equals.md]: Updated the two of these with their explicit types. This was a mindless error from writing all the sets of documentation.

File: docs/Data/Validation/instance/Equals.md (hunk 2 above)
Comment (jlmorgan) [reply docs/Data/Validation/instance/Equals.md]: "Fixed" in latest push.

File: src/FunctionalExtras/Data/Validation.cs [hunk 3]
```diff
@@ -0,0 +1,315 @@
...
+    /// <typeparam name="F">The underlying failure type.</typeparam>
+    /// <typeparam name="S">The underlying success type.</typeparam>
+    /// <param name="enumerable">The enumerable of <see cref="Validation"/></param>
+    /// <returns>The enumerable of underlying <code>Failure</code> values.</returns>
+    public static IEnumerable<F> Failures<F, S>(IEnumerable<IValidation<F, S>> enumerable) =>
+      (enumerable ?? Enumerable.Empty<IValidation<F, S>>())
+        .Where(IsNotNull)
+        .Where(IsFailure)
+        .SelectMany(FromFailure<F, S>(Enumerable.Empty<F>()))
+        .ToList();
```
Comment (JeremiahSanders) [inline src/FunctionalExtras/Data/Validation.cs]: Intentional `ToList` with `IEnumerable` return?

Comment (jlmorgan) [reply src/FunctionalExtras/Data/Validation.cs]: Can I leave it at the `SelectMany` or do I need to terminate the Linq query somehow? I sorta took this idea from the Javas where you have to close out the stream with a `.collect` call.

File: src/FunctionalExtras/Data/Validation.cs [hunk 4]
```diff
@@ -0,0 +1,315 @@
...
+    /// <typeparam name="F">The underlying failure type.</typeparam>
+    /// <typeparam name="S">The underlying succes type.</typeparam>
+    /// <param name="enumerable">The enumerable of <see cref="Validation"/></param>
+    /// <returns>The enumerable of underlying <code>Success</code> values.</returns>
+    public static IEnumerable<S> Successes<F, S>(IEnumerable<IValidation<F, S>> enumerable)
+       => (enumerable ?? Enumerable.Empty<IValidation<F, S>>())
+        .Where(IsNotNull)
+        .Where(IsSuccess)
+        .Select(FromSuccess<F, S>(default))
+        .ToList();
```
Comment (JeremiahSanders) [inline src/FunctionalExtras/Data/Validation.cs]: Intentional `ToList` with `IEnumerable` return?

File: src/FunctionalExtras.Tests/Data/ValidationTests.cs [hunk 5]
```diff
@@ -0,0 +1,707 @@
...
+          };
+          IEnumerable<string> actualResult = Validation.Failures(testEnumerable);
+
+          Assert.Equal(expectedResult, actualResult);
+        }
+      }
+
+      public class DescribeFromFailure
+      {
+        private static IEnumerable<string> _testDefaultValues = new List<string>
```
Comment (JeremiahSanders) [inline src/FunctionalExtras.Tests/Data/ValidationTests.cs]: Consider making `readonly`.

Comment (jlmorgan) [reply src/FunctionalExtras.Tests/Data/ValidationTests.cs]: Done and done. I did it in the other tests. Not sure why I missed it here.

File: src/FunctionalExtras.Tests/Data/ValidationTests.cs [hunk 6]
```diff
@@ -0,0 +1,707 @@
...
+      {
+        private static string _testFailureValue = Guid.NewGuid().ToString();
+        private static Guid _testValue = Guid.NewGuid();
+
+        [Fact]
+        public void ShouldThrowExceptionForNullPredicate()
+        {
+          Predicate<Guid> testPredicate = null;
+          
+
```
Comment (JeremiahSanders) [inline src/FunctionalExtras.Tests/Data/ValidationTests.cs]: Consider removing second newline.

Comment (jlmorgan) [reply src/FunctionalExtras.Tests/Data/ValidationTests.cs]: Done and done.

File: src/FunctionalExtras/Data/Validation.cs (hunk 3 above)
Comment (JeremiahSanders) [reply src/FunctionalExtras/Data/Validation.cs]: You can leave it with `SelectMany` unless you specifically need to execute the enumeration.

Comment (jlmorgan) [reply src/FunctionalExtras/Data/Validation.cs]: Updated this and `Either` to do the same.

File: docs/Data/Validation/instance/Equals.md [hunk 7]
```diff
@@ -0,0 +1,35 @@
+# `Validation#Equals<A, B>(object other)`
+
+Determine whether or not the `other` has the same value as the current `instance`.
```
Comment (doomchild) [inline docs/Data/Validation/instance/Equals.md]: `Determine` → `Determines`

File: docs/Data/Validation/static/Concat.md [hunk 8]
```diff
@@ -0,0 +1,53 @@
+# `Validation.Concat<F, S>(IValidation<F, S> second, IValidation<F, S> first)`
+
+Concatenates two `Failure` values together, replace a `Success` with the `Failure`; otherwise, take the first `Success`.
```
Comment (doomchild) [inline docs/Data/Validation/static/Concat.md]: `replace` → `replacing`

**Review by doomchild — CHANGES_REQUESTED** [review]
Minor formatting stuff needed.

File: docs/Data/Validation/instance/Equals.md (hunk 7 above)
Comment (jlmorgan) [reply docs/Data/Validation/instance/Equals.md]: Fixed.

File: docs/Data/Validation/static/Concat.md (hunk 8 above)
Comment (jlmorgan) [reply docs/Data/Validation/static/Concat.md]: Fixed.

**Review by doomchild — APPROVED** [review]
Seems legit.

**Pull request merged**